### PR TITLE
State Observability API REP

### PR DESCRIPTION
**Pull request by rkooo567**
Still draft now

### CODE REVIEW

File: reps/2022-04-21-state-observability-apis.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+### Shepherd of the Proposal (should be a senior committer)
+
+@edoakes
+
+## Design and Architecture
+
+### States
+“States” must be closely related and consistent to “concepts / APIs” users learn. For example, if we expose tasks as core Ray concepts, users would like to see the state of tasks. Ideally, the name of the APIs should be consistent with the concepts too.
+
+State APIs can be categorized by 3 types.
```
Comment (edoakes) [inline reps/2022-04-21-state-observability-apis.md:33]: Are there different requirements for the three categories? Why is it necessary to group them?

File: reps/2022-04-21-state-observability-apis.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+TODO(sang): Add a picture
+
+### Terminologies
+- Consumer: Consumers of state APIs. In Ray, (e.g., dashboard, CLI, and Python APIs).
+- API server: Currently a Ray dashboard. Centralized server to aggregate and provide state information of the cluster.
+- Ray agent: Currently a dashboard agent. A Python process running per node.
+- Resources: Objects created by Ray. E.g., tasks, actors, placement groups, namespace. 
+
+### Observability API architecture
```
Comment (edoakes) [inline reps/2022-04-21-state-observability-apis.md:72]: I think we should add a short summary of alternatives considered here (e.g., clients talking directly to each node).

**Review by edoakes — COMMENTED** [review]
This looks great!

@rkooo567 I think it'd be a good idea to mention that we are explicitly *not* handling storage/indexing of historical data, that can/should be done by an external system.

File: reps/2022-04-21-state-observability-apis.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+ray.summary_actors()
+ray.list_actors(state="CREATED")
+ray.get_actor(id)
+
+# REST
+/api/v1/actors/summary
+/api/v1/actors?state="CREATED"
+/api/v1/actors/{id}
+```
+
```
Comment (rkooo567) [inline reps/2022-04-21-state-observability-apis.md:114]: I am going to add a brief explanation about ray log / resources

File: reps/2022-04-21-state-observability-apis.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+- Need to expose all necessary Ray states for basic observability.
+- State APIs should be consistent / stable / well-documented. E.g., APIs names must be consistent to the concepts
+- APIs should be available to all clients, such as Dashboard, CLI, and Python APIs with the same implementation.
+- Shouldn’t introduce high overhead to the normal workloads.
+- Under no circumstances should it be possible to crash a cluster (say, up to our scalability envelope) using the observability APIs.
+
+
+### Should this change be within `ray` or outside?
+
+Accessing state is a fundamental feature to debug systems. These changes would lie within Ray to improve the basic observability of the project.
```
Comment (scv119) [inline reps/2022-04-21-state-observability-apis.md:17]: will following makes more sense here?

```
Yes, the states we want to observer belong to Ray internal components.
```

File: reps/2022-04-21-state-observability-apis.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+- Consumer: Consumers of state APIs. In Ray, (e.g., dashboard, CLI, and Python APIs).
+- API server: Currently a Ray dashboard. Centralized server to aggregate and provide state information of the cluster.
+- Ray agent: Currently a dashboard agent. A Python process running per node.
+- Resources: Objects created by Ray. E.g., tasks, actors, placement groups, namespace. 
+
+### Observability API architecture
+Currently, there are two big architectural problems.
+1. Each consumer has its own implementation for each API. E.g., ray.state.actors() and the API /logical/actors have their own implementation. each consumer in Ray uses different components/code to implement each layer, which in turn leads to having duplicated code and non-unified abstraction, which causes the unmaintainability of state APIs.
+2. No architectural pattern to guarantee stability. E.g., ray memory will just crash if the output is big.
+
```
Comment (scv119) [inline reps/2022-04-21-state-observability-apis.md:76]: is it easy to add a graph here?

**Comment by edoakes** [comment]
@rkooo567 please ping when comments addressed

**Comment by rkooo567** [comment]
@edoakes @scv119 can you have the last pass before making it an official REP?

File: reps/2022-04-21-state-observability-apis.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+**Logical states**
+- Aligned with logical concepts in which users need to learn to use Ray.
+It- Useful for beginners to advanced users.
+- E.g., Actor, Task, Objects, Resources, Placement Group, Runtime Env
+
+**Physical states**
+- States from physical components.
+- When users scale up, it will become important to observe physical states. 
+- Logs, Cluster Health, Worker processes, Nodes
```
Comment (edoakes) [inline reps/2022-04-21-state-observability-apis.md:43]:
```suggestion
- E.g., logs, cluster health, worker processes, nodes
```
nit: to be consistent w/ above and don't need to capitalize things

File: reps/2022-04-21-state-observability-apis.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+### States
+“States” must be closely related and consistent with “concepts / APIs” users learn. For example, if we expose tasks as core Ray concepts, users would like to see the state of tasks. Ideally, the name of the APIs should be consistent with the concepts too.
+
+State APIs can be categorized into 3 types.
+
+**Logical states**
+- Aligned with logical concepts in which users need to learn to use Ray.
+It- Useful for beginners to advanced users.
+- E.g., Actor, Task, Objects, Resources, Placement Group, Runtime Env
```
Comment (edoakes) [inline reps/2022-04-21-state-observability-apis.md:38]:
```suggestion
- E.g., actors, tasks, objects, resources, placement groups, runtime environments
```
nit

File: reps/2022-04-21-state-observability-apis.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+**Physical states**
+- States from physical components.
+- When users scale up, it will become important to observe physical states. 
+- Logs, Cluster Health, Worker processes, Nodes
+
+**Internal states**
+- Internal information about Ray.
+- They are mostly useful for maintainers to debug bugs, but users can also be benefited. 
+- Scheduling states, Events, Internal states of Ray components
```
Comment (edoakes) [inline reps/2022-04-21-state-observability-apis.md:48]:
```suggestion
- E.g., scheduling states, events, internal states of Ray components 
```
nit: to be consistent w/ above

**Review by edoakes — APPROVED** [review]

File: reps/2022-04-21-state-observability-apis.md (hunk 1 above)
Comment (edoakes) [reply reps/2022-04-21-state-observability-apis.md]: fine as is

**Comment by ericl** [comment]
Reminder that committer votes must be held before merging. @scv119 , shall we disable merge permissions generally so folks don't accidentally merge unapproved REPs?

**Comment by jovany-wang** [comment]
The initial motivation for addressing the state observability issues is nice to me. But I'm still concerning about the availability for large scale cluster.  
Could we elaborate the pressure and how we could scale it easily for large scale cluster? I don't think it's a good idea to degrade the performance in a large scale cluster.   


Why not use separate job or a collection of actors for aggregate/query and etc..   So that we can get good scalability on large scale cluster to support production usages, and the architecture should be clearer.

**Comment by jovany-wang** [comment]
Btw, my concerns are more about the implementation architecture, instead of the user-face APIs.

**Pull request merged**